### PR TITLE
[#4333] Pool the sliced buffers

### PR DESCRIPTION
**Pull request by ninja-**
@normanmaurer @Scottmitch please review. I have tested this on production and it does the job pretty well.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/PooledDuplicatedByteBuf.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        protected PooledDuplicatedByteBuf newObject(Handle handle) {
+            return new PooledDuplicatedByteBuf(handle);
+        }
+    };
+
+    static PooledDuplicatedByteBuf newInstance(ByteBuf buffer) {
+        PooledDuplicatedByteBuf buf = RECYCLER.get();
+        buf.init(buffer);
+        buf.setRefCnt(1);
+        buffer.retain();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledDuplicatedByteBuf.java]: can you explain me why this calls retain here ?

File: buffer/src/main/java/io/netty/buffer/PooledSlicedByteBuf.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        protected PooledSlicedByteBuf newObject(Handle handle) {
+            return new PooledSlicedByteBuf(handle);
+        }
+    };
+
+    static PooledSlicedByteBuf newInstance(ByteBuf buffer, int index, int length) {
+        PooledSlicedByteBuf buf = RECYCLER.get();
+        buf.init(buffer, index, length);
+        buf.setRefCnt(1);
+        buffer.retain();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledSlicedByteBuf.java]: same question as above .

File: buffer/src/main/java/io/netty/buffer/PooledDuplicatedByteBuf.java (hunk 1 above)
Comment (ninja-) [inline buffer/src/main/java/io/netty/buffer/PooledDuplicatedByteBuf.java]: the design is very simple - it increases parent's refcnt when created and decreases when deallocated

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledDuplicatedByteBuf.java]: ok gotcha... thanks

**Comment by normanmaurer** [comment]
@ninja- anything I wonder is if we maybe just use method names like:

```java
ByteBuf slice(boolean retain)
ByteBuf slice(int readerIndex, int length, boolean retain);
...
...
```
And if retain is `code` true we can use the pooled version and if not we will not do it. @ninja- WDYT ?

**Comment by normanmaurer** [comment]
@ninja- also please fix the check style errors

**Comment by ninja-** [comment]
yeah creating slice(boolean) was my first idea but then I switched over to rslice(). I think both of them could be fine. I'll try to get the checkstyle done today.

**Comment by normanmaurer** [comment]
@ninja- ok... then please use the slice/duplicate variants :)

**Comment by ninja-** [comment]
@normanmaurer do you mean I should keep rslice and rduplicate or something else ?

**Comment by normanmaurer** [comment]
@ninja- no... just use:
```java
duplicate(....)
slice(...)
readSlice(...)
```
Use a boolean param to identical if it should be retained or not like I said before.

**Comment by ninja-** [comment]
but I said it could be confusing as behaviour is entirely different

**Comment by normanmaurer** [comment]
@ninja- while slice(true) is 100% the same as slice().retain() is an implementation detail as for exampled the unpooled implementation would just do the same for both. The user should just ensure to call `release()` on the slice once he is done an used a boolean param.

**Comment by normanmaurer** [comment]
@ninja- I just think rslice and rduplicate are not really good names as these tell the user not much about what the impl does.

**Comment by ninja-** [comment]
so um maybe pooled{Slice,Duplicate}

**Comment by normanmaurer** [comment]
ninja having pooled in the name is strange as well as if a buffer is pooled or not is an implementation detail again.

**Comment by ninja-** [comment]
but users usually don't care about calling .release() on the normal .slices() so that's why I think .slice(boolean) would be even more confusing

**Comment by normanmaurer** [comment]
@ninja- that said if they not call release() on the slice it would not harm... it will just not return it the the Recycler and so not be able to reuse it later and thats it.

**Comment by ninja-** [comment]
no, if they don't release the pooled slice that would leak the paren buffer...

**Comment by normanmaurer** [comment]
@ninja- but only because the parent would be retained... So from the users standpoint this is identical:
```java
slice().retain();
slice(true);
```
Bot need to have `release()` called twice to not leak.

**Comment by ninja-** [comment]
um ok I think that would be fine

**Comment by normanmaurer** [comment]
@ninja- another thing we could do is not make it part of the ByteBuf interface itself but provide it as utility method of ByteBufUtils. This way the user is 100% aware of what he is doing.

Something like:
```java
public class ByteBufUtil {
    ByteBuf pooledSlice(ByteBuf buf, int readerIndex, int length);
    ByteBuf pooledDuplicate(ByteBuf buf, int readerIndex, int length);
...
}
```

**Comment by normanmaurer** [comment]
@ninja- not sure what is better though...

**Comment by ninja-** [comment]
I'd rather stick to slice(boolean)

**Comment by normanmaurer** [comment]
@ninja- ok please add it then but keep the old slice(), duplicate() etc. and also fix commit message/checkstyle

**Comment by ninja-** [comment]
ok

**Comment by normanmaurer** [comment]
@ninja- let me know once you are done. Also ensure that for the Unpooled*ByteBuf implementations it will just delegate to `slice().retain()` etc.

**Comment by normanmaurer** [comment]
@ninja- do you want to take care of this or should I submit a pr with all the changes ?

**Comment by ninja-** [comment]
@normanmaurer if you have some free time then it would be great as I don't have a lot. just don't start it from scratch.

**Comment by normanmaurer** [comment]
I think I can finish it up before the weekend. Let me take care.

2015-10-15 16:35 GMT+02:00 ninja <notifications@github.com>:

**Comment by ninja-** [comment]
cool, thanks :+1:

**Comment by ninja-** [comment]
any updates? I could slowly finish the pr if you want.

**Comment by normanmaurer** [comment]
Will post a pr tomorrow

**Comment by ninja-** [comment]
k

**Comment by normanmaurer** [comment]
Closed in favour of https://github.com/netty/netty/pull/4394

**Pull request closed without merging**